File: lib/teto_bot/interactions/whitelist.ex
defmodule TetoBot.Interactions.Whitelist do
  @moduledoc """
  Handles the /whitelist Discord slash command.
  """

  require Logger

  alias Nostrum.Struct.Interaction
  alias TetoBot.Channels
  alias TetoBot.Interactions.{Responses, Permissions}

  @spec handle_whitelist(Interaction.t()) :: :ok | Nostrum.Api.error()
  @doc """
  Handles the /whitelist command - adds a channel to the whitelist for bot operation.
  """
  def handle_whitelist(%Interaction{guild_id: guild_id, channel_id: channel_id} = interaction) do
    Permissions.with_permission(interaction, :manage_channels, fn ->
      case Channels.whitelist_channel(guild_id, channel_id) do
        {:ok, _channel} ->
          Responses.success(
            interaction,
            "Channel <##{channel_id}> whitelisted successfully!",
            ephemeral: true
          )

        {:error, error} ->
          Logger.error("Failed to whitelist channel #{channel_id}: #{inspect(error)}")
          error_message = build_error_message(error, channel_id)
          Responses.error(interaction, error_message)
      end
    end)
  end

  @spec build_error_message(any(), integer()) :: String.t()
  defp build_error_message(error, channel_id) do
    case error do
      %Ash.Error.Invalid{errors: errors} ->
        if Enum.any?(errors, &String.contains?(inspect(&1), "unique")) do
          "Channel <##{channel_id}> is already whitelisted."
        else
          "Failed to whitelist channel <##{channel_id}>. Please check the logs."
        end

      _ ->
        "Failed to whitelist channel <##{channel_id}>. Please check the logs."
    end
  end
end

File: src/commands/whitelist.ts
import {
  ChannelType,
  type ChatInputCommandInteraction,
  MessageFlags,
  SlashCommandBuilder,
} from 'discord.js'
import { Effect, Either, Runtime } from 'effect'
import { ApiService, type MainLive } from '../services'
import type { ApiError } from '../services/api/client'
import { hasManageChannelsPermission } from '../utils/permissions'

export const data = new SlashCommandBuilder()
  .setName('whitelist')
  .setDescription(
    'Whitelist a channel for the bot to operate in (requires Manage Channels permission)'
  )
  .addChannelOption((option) =>
    option
      .setName('channel')
      .setDescription('The channel to whitelist')
      .setRequired(true)
      .addChannelTypes(ChannelType.GuildText)
  )

/**
 * Build appropriate error message based on ApiError
 */
function buildErrorMessage(error: ApiError, channelId: string): string {
  // Check for "conflict" errors (409 status)
  if (error.statusCode === 409) {
    return `Channel <#${channelId}> is already whitelisted.`
  }

  return `Failed to whitelist channel <#${channelId}>: ${error.message}`
}

/**
 * Effect-based whitelist operation
 */
const whitelistChannelEffect = (
  channelId: string,
  userId: string,
  guildId: string
): Effect.Effect<void, ApiError, ApiService> =>
  Effect.gen(function* () {
    const apiService = yield* ApiService
    const effectApi = apiService.effectApi

    // Ensure user-guild relationship exists
    yield* effectApi.discord.ensureUserGuildExists({
      userId,
      guildId,
    })

    // Add channel to whitelist (create in database)
    yield* effectApi.channels.createChannel({
      channelId,
      guildId,
    })
  }).pipe(
    Effect.tapError((error) =>
      Effect.logError(
        `Failed to whitelist channel ${channelId}: ${error.message}`
      )
    )
  )

export async function execute(
  runtime: Runtime.Runtime<never>,
  live: typeof MainLive,
  interaction: ChatInputCommandInteraction
) {
  // Check permissions
  if (!hasManageChannelsPermission(interaction)) {
    await interaction.reply({
      content: 'You need the "Manage Channels" permission to use this command.',
      flags: MessageFlags.Ephemeral,
    })
    return
  }

  const channel = interaction.options.getChannel('channel', true)

  if (!interaction.guildId) {
    await interaction.reply({
      content: 'This command can only be used in a server.',
      flags: MessageFlags.Ephemeral,
    })
    return
  }

  // Convert Effect to Either and run it
  const program = whitelistChannelEffect(
    channel.id,
    interaction.user.id,
    interaction.guildId
  ).pipe(Effect.either, Effect.provide(live))
  const result = await Runtime.runPromise(runtime)(program)

  // Handle Either result
  if (Either.isLeft(result)) {
    // Error case
    const errorMessage = buildErrorMessage(result.left, channel.id)
    await interaction.reply({
      content: errorMessage,
      flags: MessageFlags.Ephemeral,
    })
  } else {
    // Success case
    await interaction.reply({
      content: `Channel <#${channel.id}> whitelisted successfully!`,
      flags: MessageFlags.Ephemeral,
    })
  }
}
